gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsBatch.java

```diff
@@ -0,0 +1,231 @@
+      this.operation = builder.operation;
+      this.options = builder.options;
+    }
+
+    private static RequestBuilder builder(Operation operation, AbstractOption... options) {
+      return new RequestBuilder(operation, options);
+    }
+
+    /**
+     * Returns the name of the zone to which the operation is applied. This field is initialized for
+     * zone create, get and delete operation, and listing DNS records and changes within a zone.
+     * Returns {@code null} in other cases.
+     */
+    public String zoneName() {
+      return zoneName;
+    }
+
+    /**
+     * Returns the id of the change request which is being retrieved. Getting a change request is
+     * the only operation when this attribute is initialized. The method returns {@code null} in the
+     * remaining cases.
+     */
+    public String changeId() {
+      return changeId;
+    }
+
+    /**
+     * Returns the change request which is being created. Creating a change request is the only
+     * operation when this attribute is initialized. The method returns {@code null} in the
+     * remaining cases.
+     */
+    public ChangeRequest changeRequest() {
+      return changeRequest;
+    }
+
+    /**
+     * Returns the zone which is being created. Creating a zone is the only operation when this
+     * attribute is initialized. The method returns {@code null} in the remaining cases.
+     */
+    public ZoneInfo zoneInfo() {
+      return zoneInfo;
+    }
+
+    /**
+     * Returns the type of the operation represented by this {@link DnsBatch.Request}. This field is
+     * always initialized.
+     */
+    public Operation operation() {
+      return operation;
+    }
+
+    /**
+     * Returns options provided to the operation. Returns an empty array if no options were
+     * provided.
+     */
+    public AbstractOption[] options() {
+      return options == null ? new AbstractOption[0] : options;
+    }
+  }
+
+  static class RequestBuilder {
+    private final AbstractOption[] options;
+    private String zoneName;
+    private String changeId;
+    private ChangeRequest changeRequest;
+    private ZoneInfo zoneInfo;
+    private final Operation operation;
+
+    RequestBuilder(Operation operation, AbstractOption... options) {
+      this.operation = operation;
+      this.options = options;
+    }
+
+    RequestBuilder zoneName(String zoneName) {
+      this.zoneName = zoneName;
+      return this;
+    }
+
+    RequestBuilder changeId(String changeId) {
+      this.changeId = changeId;
+      return this;
+    }
+
+    RequestBuilder changeRequest(ChangeRequest changeRequest) {
+      this.changeRequest = changeRequest;
+      return this;
+    }
+
+    RequestBuilder zoneInfo(ZoneInfo zoneInfo) {
+      this.zoneInfo = zoneInfo;
+      return this;
+    }
+
+    Request build() {
+      return new Request(this);
+    }
+  }
+
+  /**
+   * Represents the type of the batch operation.
+   */
+  public enum Operation {
```

Comment:
Nit: do you think it's worth adding dedicated javadoc for each enum or are the names already self-explanatory?